src/braket/default_simulator/density_matrix_simulation.py

```diff
@@ -0,0 +1,286 @@
+        dm_targets = tuple(i + qubit_count for i in targets)
+        axes = (
+            np.arange(len(targets), 2 * len(targets)),
+            dm_targets,
+        )
+        state = np.tensordot(gate_matrix, state, axes=axes)
+
+        # Arrange the index to the correct place.
+        unused_idxs = [idx for idx in range(2 * qubit_count) if idx not in dm_targets]
+        permutation = list(dm_targets) + unused_idxs
+        inverse_permutation = np.argsort(permutation)
+        state = np.transpose(state, inverse_permutation)
+
+        return state
+
+    def _apply_gate_superop(
+        state: np.ndarray, qubit_count: int, superop: np.ndarray, targets: Tuple[int, ...]
+    ) -> np.ndarray:
+        """Apply a superoperator to a density matrix
+
+        Args:
+            state (np.ndarray): initial density matrix
+            qubit_count (int): number of qubits in the circuit
+            superop (np.ndarray): superoperator to be applied to the density matrix
+            targets (Tuple[int,...]): qubits of the density matrix the superoperator applied to.
+
+        Returns:
+            state (np.ndarray): output density matrix
+        """
+        targets_new = targets + tuple([target + qubit_count for target in targets])
+
+        superop = np.reshape(superop, [2] * len(targets_new) * 2)
+        axes = (
+            np.arange(len(targets_new), 2 * len(targets_new)),
+            targets_new,
+        )
+        state = np.tensordot(superop, state, axes=axes)
+
+        # Arrange the index to the correct place.
+        unused_idxs = [idx for idx in range(2 * qubit_count) if idx not in targets_new]
+        permutation = list(targets_new) + unused_idxs
+        inverse_permutation = np.argsort(permutation)
+        state = np.transpose(state, inverse_permutation)
+
+        return state
+
+    def _apply_kraus(
+        state: np.ndarray, qubit_count: int, matrices: List[np.ndarray], targets: Tuple[int, ...]
+    ) -> np.ndarray:
+        """Apply a list of matrices {E_i} to a density matrix D according to:
+
+            .. math::
+                D \rightarrow \\sum_i E_i D E_i^{\\dagger}
+
+        Args:
+            state (np.ndarray): initial density matrix
+            qubit_count (int): number of qubits in the circuit
+            matrices (List[np.ndarray]): matrices to be applied to the density matrix
+            targets (Tuple[int,...]): qubits of the density matrix the matrices applied to.
+
+        Returns:
+            state (np.ndarray): output density matrix
+        """
+        if len(targets) > 4:
+            new_state = sum(
+                DensityMatrixSimulation._apply_gate(state, qubit_count, matrix, targets)
+                for matrix in matrices
+            )
+        else:
+            superop = sum(np.kron(matrix, matrix.conjugate()) for matrix in matrices)
+            new_state = DensityMatrixSimulation._apply_gate_superop(
+                state, qubit_count, superop, targets
+            )
```

Comment:
Why 4 instead of 3? In fact, if `len(matrices)==3`, will `_apply_gate_superop` even work?

Reply:
This concludes from the testing: https://quip-amazon.com/37dzAGOci1bw/Local-noise-simulator-benchmarking#fEQ9CAWsAoV.  Why not work?